### DISCUSSION TITLE
Using the bits parser and its error

### DISCUSSION THREAD

**Discussion by TheNeikos** [winnow-rs/winnow#498] [Q&A] [answered]
Heya!

I have a function that returns a `PResult`, but I can't easily use the `bits::bits` parser because the error traits it requires are not implemented/compatible with PResult's definition:

`bits` requires that: 
- `BitError: ParserError<(Input, usize)> + ErrorConvert<ByteError>`
- `ByteError: ParserError<Input>`

But `ErrorConvert` is only implemented for `InputError`. Since PResult is the 'main' way of using winnow, this means I have to awkwardly specify the Error type using a turbofish. `bits` has 5 generics (even if I can infer 4 of them).

So for example an invocation looks like this:

```rust
winnow::binary::bits::bits::<_, _, InputError<(_, usize)>, _, _>((
                winnow::binary::bits::pattern(0x0, 1usize),
                winnow::binary::bits::bool,
                winnow::binary::bits::bool,
                winnow::binary::bits::take(2usize)
                    .try_map(<QualityOfService as TryFrom<u8>>::try_from),
                winnow::binary::bits::bool,
                winnow::binary::bits::bool,
                winnow::binary::bits::bool,
            ))
            .parse_next(input)
            .map_err(|_: ErrMode<InputError<_>>| {
                ErrMode::from_error_kind(input, winnow::error::ErrorKind::Slice)
            })?
```

Now, am I doing something wrong? Or is this just a current limitation?

Thanks for any pointers!

**Comment by epage** [comment, marked as answer]
> But ErrorConvert is only implemented for InputError.

This is just overlooked as bit parsing doesn't get much attention (I wasn't even sure how much desire there is for this API!).

Feel free to create an issue.

> Since PResult is the 'main' way of using winnow, this means I have to awkwardly specify the Error type using a turbofish. bits has 5 generics (even if I can infer 4 of them).

Note that `PResult` only defaults the error type.  You can explicitly specify it.

**Comment by TheNeikos** [comment]
> This is just overlooked as bit parsing doesn't get much attention (I wasn't even sure how much desire there is for this API!).

I like it a lot! Its super handy when one needs it, and is definitely a step up from manually using masks.

> Feel free to create an issue.

Alright, I'll try to see if I can think of something to make it easier.

> Note that PResult only defaults the error type. You can explicitly specify it.

Oh yeah, totally! Was mostly in the context of it being unweildy in the 'default' arrangement. I'm happy that it can be fixed!

**Comment by praveenperera** [comment]
Just ran into this myself, I think docs on how to use bits with `parse_next` would be helpful
